File: 프로그래머스/2/76502. 괄호 회전하기/괄호 회전하기.py
from collections import deque 

def check(s):
    temp = []
    for c in s:
        if c == '(' or c == '[' or c == '{':
            temp.append(c)
        elif c == ')':
            if temp and temp[-1] == '(':
                del temp[-1]
            else:
                return 0
        elif c == ']':
            if temp and temp[-1] == '[':
                del temp[-1]
            else:
                return 0
        elif c == '}':
            if temp and temp[-1] == '{':
                del temp[-1]
            else:
                return 0
    if temp: return 0 # 예외
    return 1   
    
def solution(s):
    answer = 0
    s = list(s)
    for i in range(len(s)):
        s = deque(s)
        s.rotate()
        answer += check(s)
    return answer

File: 프로그래머스/2/76502. 괄호 회전하기/괄호 회전하기.java
import java.util.*;

class Solution {
    public int solution(String s) {
        int answer = 0;
        char[] chars = s.toCharArray();
        List<Character> list = new ArrayList<>();
        for(char c: chars) {
            list.add(c);
        }
        //System.out.println(list);
        for (int i = 0; i < s.length(); i++) {
            Collections.rotate(list, 1);
            answer += check(list);
        }
        
        return answer;
    }
    
    public int check(List<Character> list) {
        List<Character> temp = new ArrayList<>();
        for (char nxt: list) {
            if (nxt =='}' || nxt==']'|| nxt==')') {
                    if (temp.size() == 0) { 
                        return 0; 
                    }
                    char val = temp.get(temp.size()-1);
                    if (nxt=='}' && val=='{') {
                        temp.remove(temp.size()-1);
                        continue;
                    }
                    if (nxt==')' && val=='(') {
                        temp.remove(temp.size()-1);
                        continue;
                    }
                    if (nxt == ']' && val == '[') {
                        temp.remove(temp.size()-1);
                        continue;
                    }
                }
            else {
                temp.add(nxt);
            }
        }
        if (temp.size() != 0) {
            return 0;
        }
        return 1;
        
    }
}